common/thorhelper/hpccconfig.cpp

```diff
@@ -90,3 +90,64 @@ void getMemorySpecifications(std::unordered_map<std::string, __uint64> &memorySp
     }
     memorySpecifications["recommendedMaxMemory"] = recommendedMaxMemory;
 }
+
+static unsigned pipeProgramUpdateHookCBId = 0;
+static const char *builtInPrograms = "roxiepipe"; // csv list
+static StringBuffer allowedPipePrograms, allowedPipeProgramsWithBuiltIns;
+static CriticalSection allowedPipeCS;
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    auto updateFunc = [&](const IPropertyTree *oldComponentConfiguration, const IPropertyTree *oldGlobalConfiguration)
+    {
+        StringArray builtInList, configuredList, combinedList;
+        builtInList.appendListUniq(builtInPrograms, ",");
+
+        Owned<IPropertyTree> config = getComponentConfig();
+        // NB: containerized config supports a different format
+        if (isContainerized())
+        {
+            Owned<IPropertyTreeIterator> iter = config->getElements("allowedPipePrograms");
+            ForEach(*iter)
+                configuredList.appendUniq(iter->query().queryProp(nullptr));
+        }
+        else
+            configuredList.appendListUniq(config->queryProp("@allowedPipePrograms"), ",");
```

Comment:
I think this means old environments will default to deny all.  Should it be "*" if allowedPipePrograms is not set?  Needs a red book item if so.